internal/dataplane/parser/parser.go

```diff
@@ -386,14 +390,134 @@ func getCertFromSecret(secret *corev1.Secret) (string, string, error) {
 	return cert, key, nil
 }
 
-func getCerts(log logrus.FieldLogger, s store.Storer, secretsToSNIs map[string][]string) []kongstate.Certificate {
-	snisAdded := make(map[string]bool)
-	// map of cert public key + private key to certificate
-	type certWrapper struct {
-		cert              kong.Certificate
-		CreationTimestamp metav1.Time
+type certWrapper struct {
+	identifier        string
+	cert              kong.Certificate
+	snis              []string
+	CreationTimestamp metav1.Time
+}
+
+func getGatewayCerts(log logrus.FieldLogger, s store.Storer) []certWrapper {
+	certs := []certWrapper{}
+	gateways, err := s.ListGateways()
+	if err != nil {
+		log.WithError(err).Error("failed to list Gateways")
+		return certs
+	}
+	policies, err := s.ListReferencePolicies()
+	if err != nil {
+		log.WithError(err).Error("failed to list ReferencePolicies")
+		return certs
 	}
-	certs := make(map[string]certWrapper)
+	for _, g := range gateways {
```

Comment:
I reckon that might be a lot but can we refactor this (whole block) so that the happy paths are the least indented ones?

e.g. instead of 

```go
			if ready {
				// ...
			}
```

->

```go
			if !ready {
				continue // or use a named continue/break if in deeply nested code
			}
```

+ extract some code blocks from this loop to named `func`s so that it's easier to read? Personally I find this a bit hard to reason about.

🥇 for nice debug logs though

Reply:
Do the additional comments and newlines in 5f326535c54300407b6bf48796ec50bc921fe4f0 make this more readable? Same as https://github.com/Kong/kubernetes-ingress-controller/pull/2580/#discussion_r906293283 I don't think the internals here are that unrelated or even that complex, they just consume lots of lines because of all the debug fields an structs. I think splitting it into functions wouldn't really simplify it, since it'd be effectively the same length, but requires you to jump around to understand the full linear path.